src/gui/qgscolorbutton.cpp

```diff
@@ -462,95 +521,110 @@ void QgsColorButton::prepareMenu()
-    connect( alphaAction, &QgsColorWidgetAction::colorChanged, colorWheel, [colorWheel]( const QColor & color ) { colorWheel->setColor( color, false ); }
-           );
-    connect( colorAction, &QgsColorWidgetAction::colorChanged, alphaRamp, [alphaRamp]( const QColor & color ) { alphaRamp->setColor( color, false ); }
-           );
-    mMenu->addAction( alphaAction );
-  }
+    if ( mShowNoColorOption && mAllowOpacity )
+    {
+      QAction *noColorAction = new QAction( mNoColorString, this );
+      noColorAction->setIcon( createMenuIcon( Qt::transparent, false ) );
+      mMenu->addAction( noColorAction );
+      connect( noColorAction, &QAction::triggered, this, &QgsColorButton::setToNoColor );
+    }
 
-  if ( mColorSchemeRegistry )
-  {
-    //get schemes with ShowInColorButtonMenu flag set
-    QList< QgsColorScheme * > schemeList = mColorSchemeRegistry->schemes( QgsColorScheme::ShowInColorButtonMenu );
-    QList< QgsColorScheme * >::iterator it = schemeList.begin();
-    for ( ; it != schemeList.end(); ++it )
+    mMenu->addSeparator();
+    QgsColorWheel *colorWheel = new QgsColorWheel( mMenu );
+    colorWheel->setColor( color() );
+    QgsColorWidgetAction *colorAction = new QgsColorWidgetAction( colorWheel, mMenu, mMenu );
+    colorAction->setDismissOnColorSelection( false );
+    connect( colorAction, &QgsColorWidgetAction::colorChanged, this, &QgsColorButton::setColor );
+    mMenu->addAction( colorAction );
+    if ( mAllowOpacity )
+    {
+      QgsColorRampWidget *alphaRamp = new QgsColorRampWidget( mMenu, QgsColorWidget::Alpha, QgsColorRampWidget::Horizontal );
+      alphaRamp->setColor( color() );
+      QgsColorWidgetAction *alphaAction = new QgsColorWidgetAction( alphaRamp, mMenu, mMenu );
+      alphaAction->setDismissOnColorSelection( false );
+      connect( alphaAction, &QgsColorWidgetAction::colorChanged, this, &QgsColorButton::setColor );
+      connect( alphaAction, &QgsColorWidgetAction::colorChanged, colorWheel, [colorWheel]( const QColor & color ) { colorWheel->setColor( color, false ); }
+             );
+      connect( colorAction, &QgsColorWidgetAction::colorChanged, alphaRamp, [alphaRamp]( const QColor & color ) { alphaRamp->setColor( color, false ); }
+             );
+      mMenu->addAction( alphaAction );
+    }
+
+    if ( mColorSchemeRegistry )
     {
-      QgsColorSwatchGridAction *colorAction = new QgsColorSwatchGridAction( *it, mMenu, mContext, this );
-      colorAction->setBaseColor( mColor );
-      mMenu->addAction( colorAction );
-      connect( colorAction, &QgsColorSwatchGridAction::colorChanged, this, &QgsColorButton::setValidColor );
-      connect( colorAction, &QgsColorSwatchGridAction::colorChanged, this, &QgsColorButton::addRecentColor );
+      //get schemes with ShowInColorButtonMenu flag set
+      QList< QgsColorScheme * > schemeList = mColorSchemeRegistry->schemes( QgsColorScheme::ShowInColorButtonMenu );
```

Comment:
Should there be a special case for project color swatch that'd match the behavior of that of the data-defined property button, or do we not lock to the project color when selecting it via this swatch?

Reply:
Here's an implementation idea: add a linked( QString &name ) signal, and connect this signal to the data defined property widget in the registerLinkedWidget(...) function.

Reply:
We don't lock when selecting from swatches -- it just sets to that color. There's no way to create a linked color apart from via data defined properties/the data defined button.

Reply:
@nyalldawson , it's fine by me if we decide not to lock onto a project color if we select it via the color button swatch, I can see how that logic can be useful.